### PR TITLE
mutli-select-search component

### PR DESCRIPTION
**Pull request by garethfoote**
Initial code for baseline behaviour of multi-select-filter. 
Integration into TDR Components express app

Follows principles from News Finder front-end: [JS](https://github.com/alphagov/finder-frontend/blob/main/app/assets/javascripts/components/option-select.js) and [Rails template](https://github.com/alphagov/finder-frontend/blob/main/app/views/components/_option-select.html.erb)

### CODE REVIEW

**Comment by garethfoote** [comment]
After creating this remembered I've got tests to write. Perhaps ignore this for now.

**Comment by TomJKing** [comment]
@garethfoote You can mark the PR as "draft" if you're not quite ready for a full review.

File: src/nationalarchives/components/multi-select-search/multi-select-search.test.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    await page.keyboard.press('Tab')
+    await page.keyboard.press('Space')
+
+    const filterElement = await page.$('.js-selected-count')
+    const content = await getPropertyValue(filterElement, 'innerHTML')
+
+    await expect(content).toEqual("2 selected")
+  });
+
+  it('should always showed checked items regardless of search term', async () => {
```
Comment (TomJKing) [inline src/nationalarchives/components/multi-select-search/multi-select-search.test.ts:85]: Maybe reword test slightly: `... always show checked ...`

File: src/nationalarchives/components/multi-select-search/multi-select-search.test.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    });
+
+    // Should display 'Afar' despite not being a match for search term 'bis'
+    await expect(isHidden).toEqual(false)
+
+    const filterElement = await page.$('.js-filter-count')
+    const content = await getPropertyValue(filterElement, 'innerHTML')
+    await expect(content).toEqual("2 displayed, 1 selected")
+  });
+
```
Comment (TomJKing) [inline src/nationalarchives/components/multi-select-search/multi-select-search.test.ts:112]: Unnecessary line breaks

File: src/nationalarchives/components/multi-select-search/multi-select-search.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    // this.setupHeight();
+  };
+
+  handleKeyUp: (ev: KeyboardEvent) => void = (ev) => {
+    ev.stopPropagation();
+    if (ev.key !== "Enter") {
+      clearTimeout(Number(this.timeoutId));
+      this.timeoutId = setTimeout(this.processKeyUpTimeout, 300);
+    } else {
+      ev.preventDefault(); // prevents forms from being submitted when user presses ENTER
```
Comment (TomJKing) [inline src/nationalarchives/components/multi-select-search/multi-select-search.ts:78]: Maybe reword comment slightly: `prevents form from ...`

File: src/nationalarchives/components/multi-select-search/multi-select-search.unit.test.ts [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    it("should hide items in list", () => {
+      const root = document.createElement("div");
+      // no checked and 2 visible
+      const ul: HTMLUListElement = createListOfInputs(
+        [false, false, false, false, false],
+        [true, true, true, true, true],
+        ["Abkhazian", "Afar", "Bengali", "Bislama", "English"]
+      );
+      root.appendChild(ul);
+      // console.log(ul.children[0].textContent);
```
Comment (TomJKing) [inline src/nationalarchives/components/multi-select-search/multi-select-search.unit.test.ts:211]: Should this comment be here?

File: src/nationalarchives/components/multi-select-search/multi-select-search.unit.test.ts [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+      mss.handleKeyUp(createKeyboardEvent("b"));
+      mss.handleKeyUp(createKeyboardEvent("i"));
+      mss.handleKeyUp(createKeyboardEvent("s"));
+
+      jest.runAllTimers();
+
+      const visibleItems: number[] = mss.getVisibleItems();
+      expect(visibleItems[0]).toEqual(3);
+      // expect(setTimeout).toHaveBeenCalledTimes(3);
```
Comment (TomJKing) [inline src/nationalarchives/components/multi-select-search/multi-select-search.unit.test.ts:250]: Should this comment be here?

**Review by TomJKing — COMMENTED** [review]
Looks good. Just some minor comments.

File: src/nationalarchives/components/multi-select-search/_index.scss [hunk 6]
```diff
@@ -0,0 +1,16 @@
+.tna-multi-select-search__list-container {
```
Comment (TomJKing) [inline src/nationalarchives/components/multi-select-search/_index.scss:1]: Should the classes have `govuk` prefixed, as per the classes in the other component?

Comment (garethfoote) [reply src/nationalarchives/components/multi-select-search/_index.scss:1]: Other departments seem to create their own namespace for components. I think `govuk` is reserved for the core GDS design system. 

https://design.homeoffice.gov.uk/components?name=Pagination#Pagination-example-html
https://service-manual.nhs.uk/design-system/components/text-input

Comment (garethfoote) [reply src/nationalarchives/components/multi-select-search/_index.scss:1]: Would it be better to change the prefix to something else though? e.g. `tdr`?

File: src/nationalarchives/components/multi-select-search/multi-select-search.unit.test.ts (hunk 5 above)
Comment (garethfoote) [reply src/nationalarchives/components/multi-select-search/multi-select-search.unit.test.ts]: Removed

File: src/nationalarchives/components/multi-select-search/multi-select-search.unit.test.ts (hunk 4 above)
Comment (garethfoote) [reply src/nationalarchives/components/multi-select-search/multi-select-search.unit.test.ts]: Removed

**Review by TomJKing — DISMISSED** [review]
Looks good. Thanks

**Comment by garethfoote** [comment]
Most recent changes:
- Component configurable through HTML data params instead of function param (to reflect GDS Front End js components)
- Hiding the text input (search) until JS is enabled/inited

**Review by TomJKing — APPROVED** [review]
Looks good. Thanks
